File: TddKatasInPhp/Common/StringCalculator.php
<?php

namespace Common;

/**
 * A calculator that accepts a string input and does an operation on the provided numbers.
 */
class StringCalculator
{
    // Contains the supported number separators.
    protected $numberSeparators;
    
    // Containns the characters that can be used for content substitution.
    protected $substitutionCharacters;
    
    /**
     * Initialises a new instance on the StringCalculator class.
     */
    public function __construct()
    {
        $this->initialize();
    }
    
    /**
     * Terminates an instance on the StringCalculator class.
     */
    public function __destruct()
    {
        unset($this->numberSeparators);
        unset($this->substitutionCharacters);
    }
    
    /**
     * Adds the provided numbers together.
     * @param String $numbers The string input that contains the numbers to be added together with number separators.
     * @return Int The sum of the provided numbers.
     */
    public function add($numbers)
    {
        // Return zero if the input is null, empty string or whitespace.
        if ($numbers == NULL || $numbers === '' || trim($numbers) === '')
        {
            return 0;
        }
        
        $errorMessages = array();
        
        // Detect and set any given custom separator.
        $numbers = $this->handleCustomSeparators($numbers);
        
        // Detect any negative numbers in the input and create an error message if any are found.
        $this->handleNegativeNumbers($numbers, $errorMessages);
        
        // Detect any unregistered or unsupported separators and create an error message if any are found.
        $this->handleUnregisteresSeparators($numbers, $errorMessages);
        
        // Calculate the sum of the given numbers and detect and errors in the input.
        $sum = $this->processNumbers($numbers, $errorMessages);
        
        // Throw an exception if any errors were detected.
        if (count($errorMessages) > 0)
        {
            throw new Exception(join('\n', $errorMessages));
        }
        
        return $sum;
    }
    
    /**
     * Initialise variables.
     */
    protected function initialize()
    {
        // Initialize the default number separators.
        $this->numberSeparators = array(';', '\n');
        $this->substitutionCharacters = array
        (
            ';',
            '\'',
            '"',
            '+',
            '!',
            '-',
            '%',
            '/',
            '=',
            '~',
            '(',
            ')',
            '*',
            '_',
            ':',
            ',',
            '?',
            '.',
            '{',
            '}',
            '@',
            '&',
            '#',
            '>',
            '<',
            '[',
            ']',
            '$',
            '\\',
            '|',
            '€',
            '÷',
            '¤',
            '^',
            '§'
        );
    }
    
    /**
     * Detects any provided definition for custom number separators and sets them to be used.
     * @param String $numbers The string input that contains the numbers to be added together with number separators.
     * @return String The string input that contains the numbers to be added together without the definition of custom number separators.
     */
    protected function handleCustomSeparators($numbers)
    {
        // If any cutom separators are given in the start of the input.
        if (strpos($numbers, '//') === 0)
        {
            // Clear the default number separators.
            $this->numberSeparators = array();
            
            // Get the custom separator definition.
            $customSeparator = $this->getTheCustomSeparatorDefinitionFromTheInput($numbers);
            
            // Remove the custom separator definition from the input.
            $numbers = str_replace(sprintf('//%s\n', $customSeparator), '', $numbers);
            
            // Process the custon separator definition.
            $this->processCustomSeparatorDefinition($customSeparator);
        }
        
        return $numbers;
    }
    
    /**
     * Detects any provided negative numbers in the input.
     * @param String $numbers The string input that contains the numbers to be added together with number separators.
     * @param array $errorMessages Error messages detected when checking the input.
     */
    protected function handleNegativeNumbers($numbers, $errorMessages)
    {
        $numbers = $this->replaceSeparatorsInInput($numbers);
        $negativesFound = $this->collectNegativeNumbersFromInput($numbers);
        
        // If any negative numbers were found, create an error message.
        if (count($negativesFound) > 0)
        {
            array_push($errorMessages, sprintf('Negatives not allowed :  %s', join(', ', $negativesFound)));
        }
    }
    
    /**
     * Detects any unsupported number separators in the input.
     * @param String $numbers The string input that contains the numbers to be added together with number separators.
     * @param array $errorMessages Error messages detected when checking the input.
     */
    protected function handleUnregisteresSeparators($numbers, $errorMessages)
    {
        $originalNumbers = $numbers;
        $substitutionCharacter = $this->getSubstitutionString($numbers);
        $numbers = $this->substituteSeparatorsInInput($numbers, $substitutionCharacter);
        $numbers = $this->substituteNumbersInInput($numbers, $substitutionCharacter);
        $numbers = $this->truncateTextSubstitutionInInput($numbers, $substitutionCharacter);
        
        if ($numbers == NULL || $numbers === '' || trim($numbers) === '')
        {
            return;
        }
        
        $numberParts = explode($substitutionCharacter, $numbers);
        
        // If there's anything still remaining in the input, that's unsupported and an error should be added about it.
        foreach ($numberParts as $numberPart)
        {
            if ($numberPart == NULL || $numberPart === '' || trim($numberPart) === '')
            {
                continue;
            }
            
            $indexOfUnsupportedSeparator = strpos($originalNumbers, $numberPart);
            array_push($errorMessages, sprintf("'%s' expected but '%s' found at position %d.", join("' or '", $this->numberSeparators), $numberPart, $indexOfUnsupportedSeparator));
        }
    }
    
    /**
     * Processes the input string's numbers and returns the sum of them.
     * @param String $numbers The string input that contains the numbers to be added together with number separators.
     * @param array $errorMessages Error messages detected when checking the input.
     * @return Int The sum of the provided numbers.
     */
    protected function processNumbers($numbers, $errorMessages)
    {
        // Return zero if the input is null, empty string or whitespace.
        if ($numbers == NULL || $numbers === '' || trim($numbers) === '')
        {
            return 0;
        }
        
        $originalNumbers = $numbers;
        $substitutedNumbers = $this->substituteSeparatorsInInput($numbers, ';');
        $numbers = $this->replaceSeparatorsInInput($numbers);
        $numberParts = explode(';', $numbers);
        $sum = 0;

        // Process the numbers from the input.
        for ($index = 0; $index < count($numberParts); $index++)
        {
            $numberPart = $numberParts[$index];
            
            // If the current number is empty an error should be added.
            if ($numberPart == NULL || $numberPart === '' || trim($numberPart) === '')
            {
                // If the number is at the end of the input.
                if ($index === count($numberParts) - 1)
                {
                    array_push($errorMessages, 'Number expected but EOF found.');
                }
                else
                {
                    $indexOfUnexpected = 0;
                    
                    if ($index > 0)
                    {
                        $previousNumber = $numberParts[$index - 1];
                        $indexOfUnexpected = strpos($substitutedNumbers, $previousNumber) + strlen($previousNumber);
                        
                        foreach ($this->numberSeparators as $numberSeparator)
                        {
                            $currentVersion = sprintf("%s%s", $previousNumber, $numberSeparator);
                            
                            if (strpos($originalNumbers, $currentVersion) === true &&
                                strpos($originalNumbers, $currentVersion) === $indexOfUnexpected - strlen($previousNumber))
                            {
                                $indexOfUnexpected += strlen($numberSeparator);
                                break;
                            }
                        }
                    }
                    
                    $nextNumberIndex = strpos($substitutedNumbers, $numberParts[$index + 1]);
                    $unexpectedSeparator = substr($originalNumbers, $indexOfUnexpected, $nextNumberIndex - $indexOfUnexpected);
                    array_push($errorMessages, sprintf("Number expected but '%s' found at position %d.", $unexpectedSeparator, $indexOfUnexpected));
                }
            }
            // If the number can be parsed and are up to 1000, add it to the sum.
            else if (ctype_digit($numberPart) && intval($numberPart) <= 1000)
            {
                $sum += intval($numberPart);
            }
        }
        
        return $sum;
    }
    
    /**
     * Gets the custom delimiter definition from the input.
     * @param String $numbers The string input that contains the numbers to be added together separated by delimiters.
     * @return String The custom delimiter definition from the input.
     */
    protected function getTheCustomSeparatorDefinitionFromTheInput($numbers)
    {
        $customSeparator = substr($numbers, 2);

        if (strpos($numbers, '//[') === 0)
        {
            $customSeparator = substr($customSeparator, 0, strpos($customSeparator, ']\n') + 1);
        }
        else if (strpos($numbers, '//\n\n') === 0)
        {
            $customSeparator = '\n';
        }
        else
        {
            $customSeparator = substr($customSeparator, 0, strpos($customSeparator, '\n'));
        }
        
        return $customSeparator;
    }
    
    /**
     * Processes the custom delimiter definition from the input and adds them to the used separators.
     * @param String $customSeparator The custom delimiter definition from the input.
     */
    protected function processCustomSeparatorDefinition($customSeparator)
    {
        // If there are multiple custom separators given.
        if (strpos($customSeparator, '][') > 0)
        {
            do
            {
                $currentPart = substr($customSeparator, 0, strpos($customSeparator, '][') + 1);
                $customSeparator = str_replace($currentPart, '', $customSeparator);
                $currentPart = $this->removeSquareBracketsFromStartAndEnd($currentPart);
                array_push($this->numberSeparators, $currentPart);
            } while (strpos($customSeparator, '][') > 0);
        }

        $customSeparator = $this->removeSquareBracketsFromStartAndEnd($customSeparator);
        array_push($this->numberSeparators, $customSeparator);
        array_multisort($this->numberSeparators, SORT_STRING, SORT_DESC);
    }
    
    /**
     * Removes the square brackets from the given custom delimiter.
     * @param String $customSeparator The custom delimiter definition from the input.
     * @return String The given custom delimiter without square brackets around it.
     */
    protected function removeSquareBracketsFromStartAndEnd($customSeparator)
    {
        // If the custom separator is defined in squarely braces, remove them.
        if (strpos($customSeparator, '[') === 0 && 
            strpos($customSeparator, ']', -1) === strlen($customSeparator) - 1)
        {
            $customSeparator = substr($customSeparator, 1);
            $customSeparator = substr($customSeparator, 0, strlen($customSeparator) - 1);
        }
        
        return $customSeparator;
    }
    
    /**
     * Replaces the number separators with a single type in the input. This may shorten the length of the input.
     * @param String $numbers The string input that contains the numbers to be added together separated by delimiters.
     * @return String The input with replaced number separators.
     */
    protected function replaceSeparatorsInInput($numbers)
    {
        // Process the supported number separators.
        foreach ($this->numberSeparators as $numberSeparator)
        {
            $numbers = str_replace($numberSeparator, ';', $numbers);
        }
        
        return $numbers;
    }
    
    /**
     * Replaces the number separators with the given character in the input, but keeps the length of the input.
     * @param String $numbers The string input that contains the numbers to be added together separated by delimiters.
     * @param String $substitutionCharacter The character used for substituting a part of the input.
     * @return String The input with replaced number separators.
     */
    protected function substituteSeparatorsInInput($numbers, $substitutionCharacter)
    {
        // Process the supported number separators.
        foreach ($this->numberSeparators as $numberSeparator)
        {
            $separatorSubstitution = '';
            
            for ($i = 0; $i < strlen($numberSeparator); $i++)
            {
                $separatorSubstitution .= $substitutionCharacter;
            }
            
            $numbers = str_replace($numberSeparator, $separatorSubstitution, $numbers);
        }
        
        return $numbers;
    }
    
    /**
     * Replaces the numbers with the given character in the input, but keeps the length of the input.
     * @param String $numbers The string input that contains the numbers to be added together separated by delimiters.
     * @param String $substitutionCharacter The character used for substituting a part of the input.
     * @return String The input with replaced numbers.
     */
    protected function substituteNumbersInInput($numbers, $substitutionCharacter)
    {
        $regularExpressionPattern = "([0-9.,-]+)";
        $regexMatches = array();
        preg_match_all($regularExpressionPattern, $numbers, $regexMatches, PREG_PATTERN_ORDER);
        array_multisort($regexMatches, SORT_STRING, SORT_DESC);
        
        // Cut out each number from the input.
        foreach ($regexMatches as $regexMatch)
        {
            $separatorSubstitution = '';
            
            for ($i = 0; $i < count($regexMatch); $i++)
            {
                $separatorSubstitution .= $substitutionCharacter;
            }
            
            $numbers = str_replace($regexMatch, $separatorSubstitution, $numbers);
        }
        
        return $numbers;
    }
    
    /**
     * Truncates the given character in the input if it contains multiple occurrences in one place.
     * @param String $numbers The string input that contains the numbers to be added together separated by delimiters.
     * @param String $substitutionCharacter The character used for substituting a part of the input.
     * @return String The input with truncated occurrences of the given character.
     */
    protected function truncateTextSubstitutionInInput($numbers, $substitutionCharacter)
    {
        while (strpos($numbers, $substitutionCharacter . $substitutionCharacter) === true)
        {
            $numbers = str_replace($substitutionCharacter . $substitutionCharacter, $substitutionCharacter, $numbers);
        }
        
        return $numbers;
    }
    
    /**
     * Gets a character that is not used in the input and can be used to substitute all number delimiters.
     * @param String $numbers The string input that contains the numbers to be added together separated by delimiters.
     * @return String A character that is not used in the input and can be used to substitute all number delimiters.
     */
    protected function getSubstitutionString($numbers)
    {
        $substitutionCharacter = '';
        
        foreach ($this->substitutionCharacters as $currentCharacter)
        {
            if (strpos($numbers, $currentCharacter) === false)
            {
                $substitutionCharacter = $currentCharacter;
                break;
            }
        }
        
        return $substitutionCharacter;
    }
    
    /**
     * Collects the negative numbers from the input.
     * @param String $numbers The string input that contains the numbers to be added together separated by delimiters.
     * @return array The negative numbers from the input.
     */
    protected function collectNegativeNumbersFromInput($numbers)
    {
        $negativesFound = array();
        $regularExpressionPattern = "([0-9.,-]+)";
        $regexMatches = array();
        preg_match_all($regularExpressionPattern, $numbers, $regexMatches, PREG_PATTERN_ORDER);

        // Check each number if it's negative and gather them up.
        foreach ($regexMatches as $regexMatch)
        {
            if (ctype_digit($regexMatch) && intval($regexMatch) < 0)
            {
                array_push($negativesFound, $regexMatch);
            }
        }
        
        return $negativesFound;
    }
}

File: TddKatasInVisualCPlusPlus/Common/StringCalculator.cpp
#include "pch.h"
#include "framework.h"
#include "./StringCalculator.h"

// Initialises a new instance on the <see cref="StringCalculator"/> class.
StringCalculator::StringCalculator()
{
    this->Initialize();
}

// Terminates an instance on the <see cref="StringCalculator"/> class.
StringCalculator::~StringCalculator()
{
    this->numberSeparators.clear();
    this->substitutionCharacters.clear();
}

// Adds the provided numbers together.
int StringCalculator::Add(string numbers)
{
    // Return zero if the input is empty string or whitespace.
    if (this->isNullOrWhiteSpace(numbers))
    {
        return 0;
    }

    vector<string> errorMessages;

    // Detect and set any given custom separator.
    numbers = this->HandleCustomSeparators(numbers);

    // Detect any negative numbers in the input and create an error message if any are found.
    this->HandleNegativeNumbers(numbers, errorMessages);

    // Detect any unregistered or unsupported separators and create an error message if any are found.
    this->HandleUnregisteredSeparators(numbers, errorMessages);

    // Calculate the sum of the given numbers and detect and errors in the input.
    auto sum = this->ProcessNumbers(numbers, errorMessages);

    // Throw an exception if any errors were detected.
    if (!errorMessages.empty())
    {
        throw new exception(this->join("\n", errorMessages).c_str());
    }

    return sum;
}

// Initialise variables.
void StringCalculator::Initialize()
{
    // Initialize the default number separators.
    this->numberSeparators.push_back(";");
    this->numberSeparators.push_back("\n");
    this->substitutionCharacters.push_back(';');
    this->substitutionCharacters.push_back('\'');
    this->substitutionCharacters.push_back('"');
    this->substitutionCharacters.push_back('+');
    this->substitutionCharacters.push_back('!');
    this->substitutionCharacters.push_back('-');
    this->substitutionCharacters.push_back('%');
    this->substitutionCharacters.push_back('/');
    this->substitutionCharacters.push_back('=');
    this->substitutionCharacters.push_back('~');
    this->substitutionCharacters.push_back('(');
    this->substitutionCharacters.push_back(')');
    this->substitutionCharacters.push_back('*');
    this->substitutionCharacters.push_back('_');
    this->substitutionCharacters.push_back(':');
    this->substitutionCharacters.push_back(',');
    this->substitutionCharacters.push_back('?');
    this->substitutionCharacters.push_back('.');
    this->substitutionCharacters.push_back('{');
    this->substitutionCharacters.push_back('}');
    this->substitutionCharacters.push_back('@');
    this->substitutionCharacters.push_back('&');
    this->substitutionCharacters.push_back('#');
    this->substitutionCharacters.push_back('>');
    this->substitutionCharacters.push_back('<');
    this->substitutionCharacters.push_back('[');
    this->substitutionCharacters.push_back(']');
    this->substitutionCharacters.push_back('$');
    this->substitutionCharacters.push_back('\\');
    this->substitutionCharacters.push_back('|');
    this->substitutionCharacters.push_back('€');
    this->substitutionCharacters.push_back('÷');
    this->substitutionCharacters.push_back('¤');
    this->substitutionCharacters.push_back('^');
    this->substitutionCharacters.push_back('§');
}

// Detects any provided definition for custom number separators and sets them to be used.
string StringCalculator::HandleCustomSeparators(string numbers)
{
    // If any custom separators are given in the start of the input.
    if (this->startsWith(numbers, "//"))
    {
        // Clear the default number separators.
        this->numberSeparators.clear();

        // Get the custom separator definition.
        auto customSeparator = this->GetTheCustomSeparatorDefinitionFromTheInput(numbers);

        // Remove the custom separator definition from the input.
        this->replaceAll(numbers, "//" + customSeparator + "\n", "");

        // Process the custom separator definition.
        this->ProcessCustomSeparatorDefinition(customSeparator);
    }

    return numbers;
}

// Detects any provided negative numbers in the input.
void StringCalculator::HandleNegativeNumbers(string numbers, vector<string> errorMessages)
{
    numbers = this->ReplaceSeparatorsInInput(numbers);
    auto negativesFound = this->CollectNegativeNumbersFromInput(numbers);

    // If any negative numbers were found, create an error message.
    if (negativesFound.size() > 0)
    {
        errorMessages.push_back("Negatives not allowed : " + this->join(", ", negativesFound));
    }
}

// Detects any unsupported number separators in the input.
void StringCalculator::HandleUnregisteredSeparators(string numbers, vector<string> errorMessages)
{
    auto originalNumbers = numbers;
    auto substitutionCharacter = this->GetSubstitutionString(numbers);
    numbers = this->SubstituteSeparatorsInInput(numbers, substitutionCharacter);
    numbers = this->SubstituteNumbersInInput(numbers, substitutionCharacter);
    numbers = this->TruncateTextSubstitutionInInput(numbers, substitutionCharacter);

    if (this->isNullOrWhiteSpace(numbers))
    {
        return;
    }

    auto numberParts = this->split(numbers, string(1, substitutionCharacter));

    // If there's anything still remaining in the input, that's unsupported and an error should be added about it.
    for (auto numberPart : numberParts)
    {
        if (this->isNullOrWhiteSpace(numberPart))
        {
            continue;
        }

        auto indexOfUnsupportedSeparator = originalNumbers.find(numberPart);
        errorMessages.push_back("'" + this->join("' or '", this->numberSeparators) + "' expected but '" + numberPart + "' found at position " + to_string(indexOfUnsupportedSeparator) + ".");
    }
}

// Processes the input string's numbers and returns the sum of them.
int StringCalculator::ProcessNumbers(string numbers, vector<string> errorMessages)
{
    // Return zero if the numbers part is null, empty string or whitespace.
    if (this->isNullOrWhiteSpace(numbers))
    {
        return 0;
    }

    auto originalNumbers = numbers;
    auto substitutedNumbers = this->SubstituteSeparatorsInInput(numbers, ';');
    numbers = this->ReplaceSeparatorsInInput(numbers);
    auto numberParts = this->split(numbers, ";");
    auto sum = 0;
    int index;

    // Process the numbers from the input.
    for (index = 0; index < numberParts.size(); index++)
    {
        auto numberPart = numberParts[index];

        // If the current number is empty an error should be added.
        if (this->isNullOrWhiteSpace(numberPart))
        {
            // If the number is at the end of the input.
            if (index == numberParts.size() - 1)
            {
                errorMessages.push_back("Number expected but EOF found.");
            }
            else
            {
                auto indexOfUnexpected = 0;

                if (index > 0)
                {
                    auto previousNumber = numberParts[index - 1];
                    indexOfUnexpected = substitutedNumbers.find(previousNumber) + previousNumber.length();

                    for (auto numberSeparator : this->numberSeparators)
                    {
                        auto currentVersion = previousNumber + numberSeparator;

                        if (originalNumbers.find(currentVersion) != string::npos &&
                            originalNumbers.find(currentVersion) == indexOfUnexpected - previousNumber.length())
                        {
                            indexOfUnexpected += numberSeparator.length();
                            break;
                        }
                    }
                }

                auto nextNumberIndex = substitutedNumbers.find(numberParts[index + 1]);
                auto unexpectedSeparator = originalNumbers.substr(indexOfUnexpected, nextNumberIndex - indexOfUnexpected);
                errorMessages.push_back("Number expected but '" + unexpectedSeparator + "' found at position " + to_string(indexOfUnexpected) + ".");
            }
        }
        // If the number can be parsed and are up to 1000, add it to the sum.
        else if (this->intTryParse(numberPart) && stoi(numberPart) <= 1000)
        {
            sum += stoi(numberPart);
        }
    }

    return sum;
}

// Gets the custom delimiter definition from the input.
string StringCalculator::GetTheCustomSeparatorDefinitionFromTheInput(string numbers)
{
    auto customSeparator = numbers.substr(2);

    if (this->startsWith(numbers, "//["))
    {
        customSeparator = customSeparator.substr(0, customSeparator.find("]\n") + 1);
    }
    else if (this->startsWith(numbers, "//\n\n"))
    {
        customSeparator = "\n";
    }
    else
    {
        customSeparator = customSeparator.substr(0, customSeparator.find("\n"));
    }

    return customSeparator;
}

// Processes the custom delimiter definition from the input and adds them to the used separators.
void StringCalculator::ProcessCustomSeparatorDefinition(std::string customSeparator)
{
    // If there are multiple custom separators given.
    if (customSeparator.find("][") != string::npos)
    {
        do
        {
            auto currentPart = customSeparator.substr(0, customSeparator.find("][") + 1);
            this->replaceAll(customSeparator, currentPart, "");
            currentPart = this->RemoveSquareBracketsFromStartAndEnd(currentPart);
            this->numberSeparators.push_back(currentPart);
        } while (customSeparator.find("][") != string::npos);
    }

    customSeparator = this->RemoveSquareBracketsFromStartAndEnd(customSeparator);
    this->numberSeparators.push_back(customSeparator);
    sort(this->numberSeparators.begin(), this->numberSeparators.end(), greater<string>());
}

// Removes the square brackets from the given custom delimiter.
string StringCalculator::RemoveSquareBracketsFromStartAndEnd(string customSeparator)
{
    // If the custom separator is defined in squarely braces, remove them.
    if (this->startsWith(customSeparator, "[") &&
        this->endsWith(customSeparator, "]"))
    {
        customSeparator = customSeparator.substr(1);
        customSeparator = customSeparator.substr(0, customSeparator.length() - 1);
    }

    return customSeparator;
}

// Replaces the number separators with a single type in the input. This may shorten the length of the input.
string StringCalculator::ReplaceSeparatorsInInput(string numbers)
{
    // Process the supported number separators.
    for (auto numberSeparator : this->numberSeparators)
    {
        this->replaceAll(numbers, numberSeparator, ";");
    }

    return numbers;
}

// Replaces the number separators with the given character in the input, but keeps the length of the input.
string StringCalculator::SubstituteSeparatorsInInput(string numbers, char substitutionCharacter)
{
    // Process the supported number separators.
    for (auto numberSeparator : this->numberSeparators)
    {
        auto separatorSubstitution = "";

        for (auto i = 0; i < numberSeparator.length(); i++)
        {
            separatorSubstitution += substitutionCharacter;
        }

        this->replaceAll(numbers, numberSeparator, separatorSubstitution);
    }

    return numbers;
}

// Replaces the numbers with the given character in the input, but keeps the length of the input.
string StringCalculator::SubstituteNumbersInInput(string numbers, char substitutionCharacter)
{
    regex regularExpressionPattern("([0-9.,-]+)+");
    auto regexMatches = this->regexMatches(numbers, regularExpressionPattern);
    sort(regexMatches.begin(), regexMatches.end(), greater<string>());

    // Cut out each number from the input.
    for (auto regexMatch : regexMatches)
    {
        auto separatorSubstitution = "";

        for (auto i = 0; i < regexMatch.length(); i++)
        {
            separatorSubstitution += substitutionCharacter;
        }

        this->replaceAll(numbers, regexMatch, separatorSubstitution);
    }

    return numbers;
}

// Truncates the given character in the input if it contains multiple occurrences in one place.
string StringCalculator::TruncateTextSubstitutionInInput(string numbers, char substitutionCharacter)
{
    while (numbers.find(substitutionCharacter + substitutionCharacter) != string::npos)
    {
        this->replaceAll(numbers, string(1, substitutionCharacter) + string(1, substitutionCharacter), string(1, substitutionCharacter));
    }

    return numbers;
}

// Gets a character that is not used in the input and can be used to substitute all number delimiters.
char StringCalculator::GetSubstitutionString(string numbers)
{
    auto substitutionCharacter = (char)"";

    for (auto currentCharacter : this->substitutionCharacters)
    {
        if (numbers.find(currentCharacter) == string::npos)
        {
            substitutionCharacter = currentCharacter;
            break;
        }
    }

    return substitutionCharacter;
}

// Collects the negative numbers from the input.
vector<string> StringCalculator::CollectNegativeNumbersFromInput(string numbers)
{
    vector<string> negativesFound;
    regex regularExpressionPattern("([0-9.,-]+)+");
    auto regexMatches = this->regexMatches(numbers, regularExpressionPattern);

    // Check each number if it's negative and gather them up.
    for (auto regexMatch : regexMatches)
    {
        if (this->intTryParse(regexMatch) && stoi(regexMatch) < 0)
        {
            negativesFound.push_back(regexMatch);
        }
    }

    return negativesFound;
}

void StringCalculator::replaceAll(string& str, const string& from, const string& to)
{
    if (from.empty())
    {
        return;
    }

    auto start_pos = 0;

    while ((start_pos = str.find(from, start_pos)) != string::npos)
    {
        str.replace(start_pos, from.length(), to);
        // In case 'to' contains 'from', like replacing 'x' with 'yx'.
        start_pos += to.length();
    }
}

bool StringCalculator::endsWith(const string& str, const string& suffix)
{
    return str.size() >= suffix.size() && 0 == str.compare(str.size() - suffix.size(), suffix.size(), suffix);
}

bool StringCalculator::startsWith(const string& str, const string& prefix)
{
    return str.size() >= prefix.size() && 0 == str.compare(0, prefix.size(), prefix);
}

bool StringCalculator::isNullOrWhiteSpace(const string& str)
{
    if (str.empty() || str.find_first_not_of(' ') == string::npos)
    {
        return true;
    }

    return false;
}

string StringCalculator::join(string separator, vector<string> collection)
{
    string s;

    for (vector<string>::const_iterator i = collection.begin(); i != collection.end(); ++i)
    {
        s += *i;

        if (i != collection.end())
        {
            s += separator;
        }
    }

    return s;
}

vector<string> StringCalculator::split(const string& str, const string& delim)
{
    vector<string> tokens;
    auto prev = 0;
    auto pos = 0;

    do
    {
        pos = str.find(delim, prev);

        if (pos == string::npos)
        {
            pos = str.length();
        }

        auto token = str.substr(prev, pos - prev);

        if (!token.empty())
        {
            tokens.push_back(token);
        }

        prev = pos + delim.length();
    } while (pos < str.length() && prev < str.length());

    return tokens;
}

bool StringCalculator::intTryParse(const string& str)
{
    try
    {
        stoi(str);
        return true;
    }
    catch (const std::exception&)
    {
        return false;
    }
}

vector<string> StringCalculator::regexMatches(const string& str, const regex ptrn)
{
    vector<string> retVal;
    smatch sm;

    if (regex_search(str, sm, ptrn))
    {
        for (int i = 1; i < sm.size(); i++)
        {
            retVal.push_back(sm[i]);
        }
    }

    return retVal;
}
